### PR TITLE
[vector] Adding vector functions

### PR DESCRIPTION
**Pull request by xiangfu0**
- Support vector scalar functions, assuming vector type is stored as float[]
  - cosine distance
  - inner product
  - l2 distance
  - l1 distance
  - vector dim
  - vector norm
- Support both scalar and transform functions
- Add integration tests for v1/v2

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    double dotProduct = 0.0;
+    double norm1 = 0.0;
+    double norm2 = 0.0;
+    for (int i = 0; i < vector1.length; i++) {
+      dotProduct += vector1[i] * vector2[i];
+      norm1 += Math.pow(vector1[i], 2);
+      norm2 += Math.pow(vector2[i], 2);
+    }
+    return 1 - (dotProduct / (Math.sqrt(norm1) * Math.sqrt(norm2)));
```
Comment (KKcorps) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java:54]: Might need to check for divide by zero

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Returns the cosine distance between two vectors
+   * @param vector1 vector1
+   * @param vector2 vector2
+   * @return cosine distance
+   */
+  @ScalarFunction(names = {"cosinedistance", "cosine_distance"})
+  public static double cosineDistance(float[] vector1, float[] vector2) {
+    if (vector1.length != vector2.length) {
```
Comment (KKcorps) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java:43]: Would checking for if one vector is empty or not in addtion make sense

**Review by KKcorps — APPROVED** [review]
Only minor comments. Overall, LGTM!

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java:68]: Make `cosineDistance` can take third optional argument default value, which is used when either vector has a norm of 0

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/VectorFunctions.java]: done.

**Pull request merged**